### PR TITLE
feat: support for Vue3 JSX HMR

### PR DESCRIPTION
**Pull request by liyincode**
## Summary
support vue3 jsx hmr
## Related Links
[issue #153](https://github.com/web-infra-dev/rsbuild/issues/153)

## Checklist

- [X] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: packages/plugin-vue-jsx/package.json [hunk 1]
```diff
@@ -28,12 +28,15 @@
...
   "dependencies": {
     "@rsbuild/plugin-babel": "workspace:*",
-    "@vue/babel-plugin-jsx": "1.2.1"
+    "@vue/babel-plugin-jsx": "1.2.1",
+    "babel-plugin-vue-jsx-hmr": "1.0.0"
   },
   "devDependencies": {
     "@babel/core": "^7.23.2",
     "@rsbuild/core": "workspace:*",
+    "@rspack/core": "0.5.5",
```
Comment (chenjiahan) [inline packages/plugin-vue-jsx/package.json:37]: This seems like an unnecessary dependency

File: packages/plugin-vue-jsx/package.json [hunk 2]
```diff
@@ -28,12 +28,15 @@
...
     "@scripts/test-helper": "workspace:*",
+    "@types/babel__core": "^7.20.3",
```
Comment (chenjiahan) [inline packages/plugin-vue-jsx/package.json:39]: can we delete this

File: e2e/cases/vue/jsx-hmr/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+  expect(await page.textContent('.named')).toMatch('named 1');
+
+  // reset code
+  editFile('Comp.tsx', (code) =>
+    code.replace('default tsx updated {count', 'default tsx {count'),
+  );
+  await rsbuild.close();
+});
+
+// // not pass
```
Comment (chenjiahan) [inline e2e/cases/vue/jsx-hmr/index.test.ts:172]: We can add a comment here:

```js
TODO: see https://github.com/web-infra-dev/rsbuild/issues/153
```

File: packages/plugin-vue-jsx/src/index.ts [hunk 4]
```diff
@@ -21,10 +21,13 @@ export function pluginVueJsx(options: PluginVueJsxOptions = {}): RsbuildPlugin {
...
-            babelOptions.plugins.push([
-              require.resolve('@vue/babel-plugin-jsx'),
-              options.vueJsxOptions || {},
-            ]);
+            babelOptions.plugins.push(
+              [
+                require.resolve('@vue/babel-plugin-jsx'),
+                options.vueJsxOptions || {},
+              ],
+              [require.resolve('babel-plugin-vue-jsx-hmr')],
```
Comment (chenjiahan) [inline packages/plugin-vue-jsx/src/index.ts:29]: In some cases we do not need to add this plugin. For example, when `dev.hmr` is disabled, or when target is not `web`.

You can use the `isUsingHMR` helper from `@rsbuild/shared` to check whether HMR is enabled.

**Comment by chenjiahan** [comment]
There are some conflicts need to be resolved 😄

image

File: packages/plugin-vue-jsx/package.json (hunk 1 above)
Comment (liyincode) [reply packages/plugin-vue-jsx/package.json:37]: Indeed, I should have added it when I was testing, sorry about that

File: packages/plugin-vue-jsx/package.json (hunk 2 above)
Comment (liyincode) [reply packages/plugin-vue-jsx/package.json:39]: sure, if replaced npm package, it is useless

File: packages/plugin-vue-jsx/package.json (hunk 1 above)
Comment (chenjiahan) [reply packages/plugin-vue-jsx/package.json:37]: Haha never mind

File: packages/plugin-vue-jsx/src/index.ts (hunk 4 above)
Comment (liyincode) [reply packages/plugin-vue-jsx/src/index.ts:29]: Yes, that's right. The hmr feature is only available when dev.hmr is enabled.

**Comment by liyincode** [comment]
Please, do I need to raise a new pr after solving the above problem, I'm not too sure how to solve pr conflict😅

File: packages/plugin-vue-jsx/package.json (hunk 2 above)
Comment (liyincode) [reply packages/plugin-vue-jsx/package.json]: already removed [df85101](https://github.com/web-infra-dev/rsbuild/pull/2018/commits/df85101efcb8253cb2afd74965922e2c3c7206b1)

File: packages/plugin-vue-jsx/src/index.ts (hunk 4 above)
Comment (liyincode) [reply packages/plugin-vue-jsx/src/index.ts]: [1f85277](https://github.com/web-infra-dev/rsbuild/pull/2018/commits/1f852778b2b01c901664fb322cf68bca3987d070)
[28c1e7a](https://github.com/web-infra-dev/rsbuild/pull/2018/commits/28c1e7a2b1062cfcf80a69d76c76992202ecce50)

**Comment by liyincode** [comment]
> There are some conflicts need to be resolved 😄
> 
> image

already resolved

**Comment by liyincode** [comment]
Hi, I forgot the run steps in the `CONTRIBUTING.md` before, I've followed the steps now and it should be fine.

But, After I merged the main branch, I ran `pnpm run e2e`, and found that one of the test cases didn't pass😅

image

**Comment by liyincode** [comment]
> > Hi, I forgot the run steps in the `CONTRIBUTING.md` before, I've followed the steps now and it should be fine.
> > But, After I merged the main branch, I ran `pnpm run e2e`, and found that one of the test cases didn't pass😅
> 
> @liyincode did you run `build` again after you merge main branch?

oh, i didn't. I have now executed build and e2e test has passed👍

i exected `pnpm run lint`, is this passed?
image

**Comment by chenjiahan** [comment]
> i exected pnpm run lint, is this passed?

It seems that lint is passed. When executing ls-lint locally, the result is abnormal. You can ignore it and we will handle it.

**Comment by liyincode** [comment]
Only windows didn't pass, I found two issues in the e2e test. The first one seems to need your help, and I will solve the second one on a Windows machine.
image
image

**Comment by chenjiahan** [comment]
You can ignore the first `decoratorBeforeExport` error, it is expected 😄

**Comment by liyincode** [comment]
Hi, I'm not sure what specifically causes the issue where e2e tests fail only on Windows platform. 

Everything works fine on my Mac using a Windows 11 virtual machine, but the problem recurs on my home desktop with Windows 11. I suspect it might be related to the ARM vs x86 architecture or execution speed, but I'm unsure how to fix it. 

Interestingly, when I test single e2e unit `run npx playwright test cases/vue/jsx-hmr/index.test.ts -g "hmr: named export via specifier" `within the e2e directory, everything works fine. 

However, running all tests with `npx playwright test cases/vue/jsx-hmr/index.test.ts` reproduces the issue. I suspected it might be due to launching rsbuild run dev multiple times.

Surprisingly, by using a single `rsbuild` and `page` instance with `test.describe`, I found that the problem was resolved, and everything worked normally.

**Comment by chenjiahan** [comment]
HMR related cases sometimes will fail in Windows, you can add a condition to skip that as a workaround:

image

There should be a more fundamental reason behind this, and we need to take time to investigate.

**Comment by liyincode** [comment]
ok, i has disabled the hmr e2e test in windows, but there's another error I need to fix.🥲

**Review by chenjiahan — APPROVED** [review]

**Comment by chenjiahan** [comment]
Thanks for your great work! 👍👍  It will be released in Rsbuild v0.6.0

**Comment by liyincode** [comment]
I'm confused about the difference between these three test types
- test
- rspackOnlyTest
- webpackOnlyTest

I'm using rspackOnlyTest to write e2e test cases, and using `pnpm e2e:rspack -g vue` `pnpm e2e:webpack -g vue` doesn't report an error


If I write an e2e test case using test, using `pnpm e2e:rspack -g vue` doesn't give me an error, but using `pnpm e2e:webpack -g vue` does!

**Comment by liyincode** [comment]
> I'm confused about the difference between these three test types
> 
> * test
> * rspackOnlyTest
> * webpackOnlyTest
> 
> I'm using rspackOnlyTest to write e2e test cases, and using `pnpm e2e:rspack -g vue` `pnpm e2e:webpack -g vue` doesn't report an error
> 
> If I write an e2e test case using test, using `pnpm e2e:rspack -g vue` doesn't give me an error, but using `pnpm e2e:webpack -g vue` does! image

I now mix the use of `rspackOnlyTest` and `test` in this way, I don't know if there will be any problems.
image

**Comment by chenjiahan** [comment]
That's ok. Some of the E2E tests in Rsbuild are run by both Rspack and Webpack at the same time. This is to check that the functionality of Rspack is correctly aligned with Webpack.

For the Vue related E2E cases, we only need to support Rspack, so we can use `rspackOnlyTest` in Vue E2E cases. `rspackOnlyTest` is a function based on the `test` method and can be mixed with test.

**Comment by liyincode** [comment]
> Thanks for your great work! 👍👍 It will be released in Rsbuild v0.6.0

Hahaha, thanks for the outstanding efforts of the rsbuild team, I'm very happy to be able to contribute a little.😃

**Pull request merged**